tech_reports/LLMs/llms.md

```diff
@@ -57,12 +57,189 @@ Authors:
+```
+
+#### ReduceScatter
+The ReduceScatter operation reduces the data across all devices and shards the result of the reduction over a specified dimension across all devices.
+
+- Supported Topologies: Ring, Linear
+- Supported number of links: 1
+- Arguments
+  - mesh_tensor: a tensor mapped to a mesh_device via mesh_mapper
+  - dim: the dimension to concatenate
+  - cluster_axis: cluster axis to gather along
+  - num_links: number of ethernet links to be used
+  - topology: topology configuration ttnn.Ring or ttn.Linear
+
+*Figure: Example usage of Ring Reduce-Scatter on 2x4 mesh_device*
+
+```py
+# Execute Reduce-Scatter on the sharded tensor
+# Assuming mesh_tensor_sharded is a sharded tensor over 8 devices where each devices contains a 32x32 sized chunk of data, the output_tensor is again of size 32x32 on each devices but reduced over all devices
+output_tensor = ttnn.reduce_scatter(mesh_tensor_sharded, dim=3, num_links=1)
+```
+
+*Figure: Example usage of Linear Reduce-Scatter on 2x4 mesh_device*
+
+```py
+# Execute Reduce-Scatter on the sharded tensor
+# Assuming mesh_tensor_sharded is a sharded tensor over 2x4 devices where each devices contains a 32x32 sized chunk of data, the output_tensor is of size 32x32 on each device but reduces over each row of devices
+output_tensor = ttnn.reduce_scatter(mesh_tensor_sharded, dim=3, num_links=1, cluster_axis=1, mesh_device=mesh_device, topology=ttnn.Topology.Linear)
+```
+
+#### AllReduce
+The AllReduce operation reduces data across all devices and stores the entire tensor on each device (replication). It is performed using an AllGather followed by a ReduceScatter.
+
+A fused version of AllReduce is planned, but currently only the composite of AllGather+ReduceScatter is supported.
+
+#### Sharding schemes for decode
+In decode mode, activations are generally stored in L1 memory, while weights, which are too large, need to be stored in DRAM. The main bottleneck in decode mode is thereby DRAM bandwidth required to load model weights.
+
+The activations in decode mode are so small because they contain the batch size (=users) in the height dimension while sequence length is 1. 
+The only exception is the attention operations computing `softmax(Q*KˆT)*V`. The activation width is the model dim (e.g. 8192 for Llama3-70b). 
+Activations are not sharded in the height dimension; however, depending on the operation and model, they may be sharded in the width dimension.
+
+Matmul weights on the other hand can be sharded in width, height or both. Sharding weights across multiple devices significantly reduces DRAM pressure per device, resulting in notable latency improvements. Below is a summary of useful sharding schemes for sharding weights in decode mode. Which scheme to use will depend on the shape and size of the model weights and the target device topology.
+
+##### **1D Column parallel**
+**When to use:** 1D cluster topologies
```

Comment:
This "When to use" headers are confusing to me. Maybe we can remove it, it is obvious we are describing 1D/2D sharding.

Reply:
removed them!